intro_to_ada.rst

```diff
@@ -1904,23 +1904,747 @@ Abstract data types
+
+    begin
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+       New_Line;
+       Put_Line ("Reversing My_Color...");
+       New_Line;
+       Reverse_Color_Array (My_Colors);
+
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+    end Test_Non_Generic_Reverse_Colors;
+
+The procedure ``Reverse_Color_Array`` takes an array of colors and starts
+by swapping the first and last elements of the array, and continues doing
+that with the next elements until the middle of array. At this point, the
+whole array has been reversed, as we can see in the text output of the
+test application.
+
+In order to abstract this procedure, we will declare formal types for
+three components of the algorithm:
+
+    - the elements of the array (``Color`` type in the example)
+
+    - the range used for the array (``Integer`` range in the example)
+
+    - the actual array type (``Color_Array`` type in the example)
+
+This is a generic version of the algorithm:
+
+.. code-block:: ada
+
+    with Ada.Text_IO;
+    use  Ada.Text_IO;
+
+    procedure Test_Reverse_Colors is
+       generic
+          type T is private;
+          type Index is range <>;
+          type Array_T is array (Index range <>) of T;
+       procedure Generic_Reverse_Array (X : in out Array_T);
+
+       procedure Generic_Reverse_Array (X : in out Array_T) is
+       begin
+          for I in X'First .. (X'Last + X'First) / 2 loop
+             declare
+                Tmp     : T;
+                X_Left  : T renames X (I);
+                X_Right : T renames X (X'Last + X'First - I);
+             begin
+                Tmp     := X_Left;
+                X_Left  := X_Right;
+                X_Right := Tmp;
+             end;
+          end loop;
+       end Generic_Reverse_Array;
+
+       type Color is (Black, Red, Green, Blue, White);
+       type Color_Array is array (Integer range <>) of Color;
+
+       procedure Reverse_Color_Array is new
+         Generic_Reverse_Array (T => Color, Index => Integer, Array_T => Color_Array);
+
+       My_Colors : Color_Array (1 .. 5) := (Black, Red, Green, Blue, White);
+
+    begin
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+       New_Line;
+       Put_Line ("Reversing My_Color...");
+       New_Line;
+       Reverse_Color_Array (My_Colors);
+
+       for C of My_Colors loop
+          Put_Line ("My_Color: " & Color'Image (C));
+       end loop;
+
+    end Test_Reverse_Colors;
+
+As mentioned above, we're abstracting three components of the algorithm:
+
+    - the ``T`` type abstracts the elements of the array
+
+    - the ``Index`` type abstracts the range used for the array
+
+    - the ``Array_T`` type abstracts the array type and makes use of the
+      formal declarations of the ``T`` and ``Index`` types.
+
+
+Abstracting definitions into packages
```

Comment:
I think this might be beyond the scope of our intro to Ada class (generic formal packages), if only because that style of programming is not very common in Ada.